nomad/node_pool_endpoint_test.go

```diff
@@ -0,0 +1,1112 @@
+			"env":   "prod",
+			"index": "1",
+		},
+	}
+	err := s.fsm.State().UpsertNodePools(structs.MsgTypeTestSetup, 1000, []*structs.NodePool{
+		poolDev1,
+		poolDev2,
+		poolProd1,
+	})
+	must.NoError(t, err)
+
+	// Create test ACL tokens.
+	devToken := mock.CreatePolicyAndToken(t, s.fsm.State(), 1001, "dev-node-pools",
+		mock.NodePoolPolicy("dev-*", "read", nil),
+	)
+	prodToken := mock.CreatePolicyAndToken(t, s.fsm.State(), 1003, "prod-node-pools",
+		mock.NodePoolPolicy("prod-*", "read", nil),
+	)
+	noPolicyToken := mock.CreateToken(t, s.fsm.State(), 1005, nil)
+	allPoolsToken := mock.CreatePolicyAndToken(t, s.fsm.State(), 1007, "all-node-pools",
+		mock.NodePoolPolicy("*", "read", nil),
+	)
+
+	testCases := []struct {
+		name     string
+		token    string
+		expected []string
+	}{
+		{
+			name:  "management token lists all",
+			token: root.SecretID,
+			expected: []string{
+				"all", "default",
+				"dev-1", "dev-2", "prod-1",
+			},
+		},
+		{
+			name:     "dev token lists dev",
+			token:    devToken.SecretID,
+			expected: []string{"dev-1", "dev-2"},
+		},
+		{
+			name:     "prod token lists prod",
+			token:    prodToken.SecretID,
+			expected: []string{"prod-1"},
+		},
+		{
+			name:  "all pools token lists all",
+			token: allPoolsToken.SecretID,
+			expected: []string{
+				"all", "default",
+				"dev-1", "dev-2", "prod-1",
+			},
+		},
+		{
+			name:     "no policy token",
+			token:    noPolicyToken.SecretID,
+			expected: []string{},
+		},
+		{
+			name:     "no token",
+			token:    "",
+			expected: []string{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Make node pool list request.
+			req := &structs.NodePoolListRequest{
+				QueryOptions: structs.QueryOptions{
+					Region:    "global",
+					AuthToken: tc.token,
+				},
+			}
+			var resp structs.NodePoolListResponse
+			err := msgpackrpc.CallWithCodec(codec, "NodePool.List", req, &resp)
+			must.NoError(t, err)
+
+			// Check response.
+			got := make([]string, len(resp.NodePools))
+			for i, pool := range resp.NodePools {
+				got[i] = pool.Name
+			}
+			must.Eq(t, tc.expected, got)
+		})
+	}
+}
+
+func TestNodePoolEndpoint_List_BlockingQuery(t *testing.T) {
+	ci.Parallel(t)
+
+	s, cleanupS := TestServer(t, nil)
+	defer cleanupS()
+
+	codec := rpcClient(t, s)
+	testutil.WaitForLeader(t, s.RPC)
+
+	// Populate state with some node pools.
+	// Insert triggers watchers.
+	pool := mock.NodePool()
+	time.AfterFunc(100*time.Millisecond, func() {
+		s.fsm.State().UpsertNodePools(structs.MsgTypeTestSetup, 1000, []*structs.NodePool{pool})
+	})
+
+	req := &structs.NodePoolListRequest{
+		QueryOptions: structs.QueryOptions{
+			Region:        "global",
+			MinQueryIndex: 1,
+		},
+	}
+	start := time.Now()
+	var resp structs.NodePoolListResponse
+	err := msgpackrpc.CallWithCodec(codec, "NodePool.List", req, &resp)
+	must.NoError(t, err)
+
+	elapsed := time.Since(start)
+	must.Greater(t, 100*time.Millisecond, elapsed)
```

Comment:
Isn't this racy? The 100ms `UpsertNodePools` timer has already started by the time we call `start := time.Now()`, so it's possible to fall under this if the blocking query returns quickly enough after the watch fires. I don't think the specific timing is actually all that relevant to the test as much as the index is?

Reply:
Hum...true. I guess the time checks for the request being blocked for a while, but if it doesn't than the index check fails anyway. This was mostly copy-past from other similar tests so I didn't think too much about them 😅